scheduler/cloudformation/cloudformation.go

```diff
@@ -280,13 +294,118 @@ func (s *Scheduler) submit(ctx context.Context, tx *sql.Tx, app *scheduler.App,
 	return nil
 }
 
-// updateStack performs a stack update, but waits for the stack to enter a
-// stable state before starting.
+func (s *Scheduler) createStack(input *cloudformation.CreateStackInput, done chan error) error {
+	p := func() error {
+		_, err := s.cloudformation.CreateStack(input)
+		return err
+	}
+	waiter := s.cloudformation.WaitUntilStackCreateComplete
+	locked := make(chan struct{})
+	return s.performStackOperation(*input.StackName, p, waiter, locked, done)
+}
+
+// updateStack enqueues the stack update, and waits a set amount of time for the
+// stack update to be submitted.
+func (s *Scheduler) updateStack(input *cloudformation.UpdateStackInput, done chan error) error {
+	locked := make(chan struct{})
+	errCh := make(chan error)
+
+	go func() {
+		p := func() error { return s.executeStackUpdate(input) }
+		waiter := s.cloudformation.WaitUntilStackUpdateComplete
+		errCh <- s.performStackOperation(*input.StackName, p, waiter, locked, done)
+	}()
+
+	var err error
+	select {
+	case <-time.After(lockWait):
+		// FIXME: At this point, we don't want to affect UX by waiting
+		// around, so we return. But, if the stack update times out, or
+		// there's an error, that information is essentially silenced.
+		return nil
+	case err = <-errCh:
+	case <-locked:
+		// if a lock is obtained within the time frame, we might as well
+		// just wait for the update to get submitted.
+		err = <-errCh
+	}
+
+	return err
+}
+
+// performStackOperation enqueues the stack update/create. This function returns as soon
+// as the stack update/create has been submitted.
 //
-// TODO: Timeout?
-func (s *Scheduler) updateStack(stack *cloudformation.Stack, input *cloudformation.UpdateStackInput, wait bool) (*cloudformation.UpdateStackOutput, error) {
-	status := *stack.StackStatus
-	stackName := input.StackName
+// * If there are no operations currently in progress, the stack operation will execute.
+// * If there is a currently active stack operation, the operation will be queued behind it.
+//
+// An error will be sent on the `done` channel when the stack operation completes.
+func (s *Scheduler) performStackOperation(stackName string, fn func() error, waiter func(*cloudformation.DescribeStacksInput) error, locked chan struct{}, done chan error) error {
+	l, err := newAdvisoryLock(s.db, stackName)
+	if err != nil {
+		return err
+	}
+
+	// Cancel any pending stack operation, since this one obsoletes older
+	// operations.
+	if err := l.CancelPending(); err != nil {
+		return fmt.Errorf("error canceling pending stack operation: %v", err)
+	}
+
+	if err := l.Lock(); err != nil {
+		// This will happen when a newer stack update obsoletes
+		// this one. We simply return nil.
+		//
+		// TODO: Should we return an error here?
+		if err == pglock.Canceled {
+			return nil
```

Comment:
Yep, I think this ties into https://github.com/remind101/empire/pull/852#discussion_r65461995